hw/xfree86/common/xf86platformBus.c

```diff
@@ -302,70 +301,82 @@ xf86platformProbe(void)
                             LogMessageVerb(X_CONFIG, 1, "OutputClass \"%s\" ModulePath for driver %s overridden with \"%s\"\n",
                                     cl->identifier, cl->driver, driver_path);
                         } else {
-                            XNFasprintf(&driver_path, "%s", xf86ModulePath);
+                            snprintf(driver_path, sizeof(driver_path)-1, "%s", xf86ModulePath);
                             LogMessageVerb(X_CONFIG, 1, "OutputClass \"%s\" ModulePath for driver %s reset to standard \"%s\"\n",
                                     cl->identifier, cl->driver, driver_path);
                         }
                     } else {
-                        driver_path = NULL;
                         LogMessageVerb(X_CONFIG, 1, "OutputClass \"%s\" ModulePath for driver %s reset to default\n",
                                 cl->identifier, cl->driver);
                     }
                     if (*(cl->driver)) LoaderSetPath(cl->driver, driver_path);
                     if (cl->modules) {
                         LogMessageVerb(X_CONFIG, 1, "    and for modules \"%s\" as well\n",
                                 cl->modules);
-                        XNFasprintf(&copy, "%s", cl->modules);
-                        curr = copy;
+                        char *copy = strdup(cl->modules ? cl->modules : "");
+                        char *curr = copy;
+                        char *next;
                         while ((curr = strtok_r(curr, ",", &next))) {
                             if (*curr) LoaderSetPath(curr, driver_path);
                             curr = NULL;
                         }
                         free(copy);
                     }
-                    free(driver_path);
                 }
                 else if (cl->modules) {
+                    char driver_path[PATH_MAX] = { 0 };
                     if (cl->modulepath) {
                         if (*(cl->modulepath)) {
-                            XNFasprintf(&driver_path, "%s,%s", cl->modulepath, xf86ModulePath);
+                            snprintf(driver_path, sizeof(driver_path)-1, "%s,%s", cl->modulepath, xf86ModulePath);
                             LogMessageVerb(X_CONFIG, 1, "OutputClass \"%s\" ModulePath for modules %s overridden with \"%s\"\n",
                                     cl->identifier, cl->modules, driver_path);
                         } else {
-                            XNFasprintf(&driver_path, "%s", xf86ModulePath);
+                            snprintf(driver_path, sizeof(driver_path)-1, "%s", xf86ModulePath);
                             LogMessageVerb(X_CONFIG, 1, "OutputClass \"%s\" ModulePath for modules %s reset to standard \"%s\"\n",
                                     cl->identifier, cl->modules, driver_path);
                         }
                     } else {
-                        driver_path = NULL;
```

Comment:
There’s a `LoaderSetPath` call below and non-null `driver_path` can trigger slightly different behaviour. We have to be careful here.